### DISCUSSION TITLE
Clarification about reusing ch.Client in different goroutines

### DISCUSSION THREAD

**Discussion by develar** [ClickHouse/ch-go#80] [General]
Could you please clarify — is `ch.Client` thread safe, is it possible perform `Do` in several goroutines in the same time?

**Comment by ernado** [comment]
No, it is not goruoutine-safe.

You can't call `Do` or `Ping` concurrently.

**Comment by develar** [comment]
Thanks. I implemented pool using https://github.com/jackc/puddle (also support generics). Another issue, is that `ch` doesn't reconnect client automatically in case of closed connection (due to inactivity).

**Reply by ernado** [reply]
No auto-reconnect is implemented.

Currently client is pretty low level, but I have plans to create high level features like std sql compatibility, reflection-based mapping, generics for complex types (if possible, ArrayOf should already work), pooling, reconnects, etc.

**Comment by pmrt** [comment]
Could you add a big warning in the readme with this? It's looking great and I will probably switch to this in the future but low-level or not I think that reconnect, pooling and thread safe connections are essential features for db clients that most users will take for granted if it is not specified in the readme.

Keep up the good work, we're looking forward to it!

**Reply by ernado** [reply]
Of course, will clarify this in readme soon, thanks!

**Comment by 0x19** [comment]
@ernado In order for me to fully understand the problem, even if it's separated package, theoretically, what would need to be done for Do to support concurrency? Pool acquire and release will allow not using mutexes on the Do but it adds overhead of connection establishment. 

Would like to understand problem from your perspective. Perhaps I can help here?
